File: crates/escurel-md/src/wikilink.rs
//! Typed wikilink parser.
//!
//! Escurel pages cite each other through wikilinks of the form
//!
//! ```text
//! [[skill::id#anchor@version|alias]]
//! ```
//!
//! All segments except the link target are optional. A bare
//! `[[id]]` is interpreted as an id-only reference; the resolver
//! (in `escurel-index`) decides at lookup time whether that id
//! names a skill or an instance.
//!
//! ## Why regex and not a markdown AST
//!
//! The Escurel storage spec is explicit about this:
//!
//! > Parse wikilinks using the regex-plus-code-region-stripping
//! > parser (do **not** use a markdown AST library — they
//! > fragment text on `[`).
//!
//! Markdown ASTs treat `[` as the start of a link and split runs
//! of text whenever they see one, which destroys the
//! `[[skill::id]]` token before we ever see it. We strip fenced
//! and inline code regions ourselves and run a regex over what
//! remains.

use std::sync::LazyLock;

use regex::Regex;

/// One wikilink occurrence, decomposed.
///
/// Shape mirrors `WikilinkParsed` in `docs/spec/protocol.md`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WikilinkParsed {
    /// `skill` segment of a typed `[[skill::id]]`. `None` for a
    /// bare `[[id]]`.
    pub skill: Option<String>,
    /// `id` segment. `None` only when the input was malformed (a
    /// typed link like `[[skill::]]` with nothing after `::`).
    pub id: Option<String>,
    /// `#anchor` segment, if present.
    pub anchor: Option<String>,
    /// `@version` segment, if present.
    pub version: Option<String>,
    /// `|alias` (display text) segment, if present.
    pub alias: Option<String>,
}

/// Parse every wikilink in `markdown`, skipping any wikilink that
/// sits inside a fenced code block (` ``` `) or an inline code
/// span (`` `…` ``).
///
/// Returns links in document order.
pub fn parse_wikilinks(markdown: &str) -> Vec<WikilinkParsed> {
    let stripped = strip_code_regions(markdown);
    WIKILINK_RE
        .captures_iter(&stripped)
        .filter_map(|cap| parse_one(&cap[1]))
        .collect()
}

/// Parse the contents between `[[` and `]]` into a [`WikilinkParsed`].
/// Returns `None` if the content can't yield even an id or a skill
/// (e.g. only whitespace).
fn parse_one(content: &str) -> Option<WikilinkParsed> {
    let (target, alias) = split_first(content, '|');
    let (target, version) = split_first(target, '@');
    let (target, anchor) = split_first(target, '#');
    let (skill, id) = match target.split_once("::") {
        Some((s, i)) => (some_if_nonempty(s), some_if_nonempty(i)),
        None => (None, some_if_nonempty(target)),
    };

    if skill.is_none() && id.is_none() {
        return None;
    }

    Some(WikilinkParsed {
        skill,
        id,
        anchor: anchor.and_then(some_if_nonempty),
        version: version.and_then(some_if_nonempty),
        alias: alias.and_then(some_if_nonempty),
    })
}

/// Split on the first occurrence of `sep`. Trims surrounding
/// whitespace on both halves.
fn split_first(s: &str, sep: char) -> (&str, Option<&str>) {
    match s.split_once(sep) {
        Some((l, r)) => (l.trim(), Some(r.trim())),
        None => (s.trim(), None),
    }
}

fn some_if_nonempty(s: &str) -> Option<String> {
    if s.is_empty() {
        None
    } else {
        Some(s.to_owned())
    }
}

/// Replace fenced code blocks and inline code spans with
/// equal-length whitespace, preserving newlines and byte offsets
/// so the wikilink regex never sees them.
fn strip_code_regions(input: &str) -> String {
    // Pass 1: fenced code blocks. A line whose trimmed-leading
    // content begins with ``` toggles in/out of a fence and is
    // itself blanked (so a ``` line never matches a wikilink).
    let mut out = String::with_capacity(input.len());
    let mut in_fence = false;
    for line in input.split_inclusive('\n') {
        let is_fence_marker = line.trim_start().starts_with("```");
        if is_fence_marker || in_fence {
            push_blanked(&mut out, line);
            if is_fence_marker {
                in_fence = !in_fence;
            }
        } else {
            out.push_str(line);
        }
    }

    // Pass 2: inline code spans on what's left.
    INLINE_CODE_RE
        .replace_all(&out, |caps: &regex::Captures| " ".repeat(caps[0].len()))
        .into_owned()
}

fn push_blanked(out: &mut String, line: &str) {
    for ch in line.chars() {
        out.push(if ch == '\n' { '\n' } else { ' ' });
    }
}

static WIKILINK_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\[\[([^\[\]\r\n]+?)\]\]").expect("wikilink regex compiles"));

static INLINE_CODE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"`[^`\n]*`").expect("inline code regex compiles"));

File: packages/escurel_explorer_kit/lib/md/wikilink.dart
/// Dart mirror of `escurel-md`'s typed wikilink parser.
///
/// Grammar: `[[skill::id#anchor@version|alias]]`. All segments
/// except the link target are optional; a bare `[[id]]` is an
/// id-only reference. Fenced code blocks (` ``` `) and inline code
/// spans (`` ` ``) are stripped before scanning.
library;

/// One wikilink occurrence, decomposed.
class WikilinkRef {
  const WikilinkRef({
    this.skill,
    this.id,
    this.anchor,
    this.version,
    this.alias,
  });

  final String? skill;
  final String? id;
  final String? anchor;
  final String? version;
  final String? alias;

  /// Render back to wikilink source for round-trip use.
  String toMarkup() {
    final target = [
      if (skill != null) '$skill::',
      if (id != null) id,
    ].whereType<String>().join();
    final suffixes = [
      if (anchor != null) '#$anchor',
      if (version != null) '@$version',
      if (alias != null) '|$alias',
    ].join();
    return '[[$target$suffixes]]';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is WikilinkRef &&
          skill == other.skill &&
          id == other.id &&
          anchor == other.anchor &&
          version == other.version &&
          alias == other.alias;

  @override
  int get hashCode => Object.hash(skill, id, anchor, version, alias);

  @override
  String toString() =>
      'WikilinkRef(skill: $skill, id: $id, anchor: $anchor, version: $version, alias: $alias)';
}

final RegExp _wikilinkRe = RegExp(r'\[\[([^\[\]\r\n]+?)\]\]');
final RegExp _inlineCodeRe = RegExp(r'`[^`\n]*`');

/// Parse every wikilink in [markdown] in document order, skipping
/// wikilinks inside fenced code blocks or inline code spans.
List<WikilinkRef> parseWikilinks(String markdown) {
  final stripped = _stripCodeRegions(markdown);
  return _wikilinkRe
      .allMatches(stripped)
      .map((m) => _parseOne(m.group(1)!))
      .whereType<WikilinkRef>()
      .toList(growable: false);
}

WikilinkRef? _parseOne(String content) {
  final (target1, alias) = _splitFirst(content, '|');
  final (target2, version) = _splitFirst(target1, '@');
  final (target3, anchor) = _splitFirst(target2, '#');

  String? skill;
  String? id;
  final sepIdx = target3.indexOf('::');
  if (sepIdx >= 0) {
    skill = _someIfNonempty(target3.substring(0, sepIdx).trim());
    id = _someIfNonempty(target3.substring(sepIdx + 2).trim());
  } else {
    id = _someIfNonempty(target3.trim());
  }

  if (skill == null && id == null) return null;

  return WikilinkRef(
    skill: skill,
    id: id,
    anchor: anchor != null ? _someIfNonempty(anchor.trim()) : null,
    version: version != null ? _someIfNonempty(version.trim()) : null,
    alias: alias != null ? _someIfNonempty(alias.trim()) : null,
  );
}

(String, String?) _splitFirst(String s, String sep) {
  final idx = s.indexOf(sep);
  if (idx < 0) return (s.trim(), null);
  return (s.substring(0, idx).trim(), s.substring(idx + 1).trim());
}

String? _someIfNonempty(String s) => s.isEmpty ? null : s;

String _stripCodeRegions(String input) {
  final blanked = StringBuffer();
  var inFence = false;
  for (final line in _linesInclusive(input)) {
    final isFenceMarker = line.trimLeft().startsWith('```');
    if (isFenceMarker || inFence) {
      for (final unit in line.runes) {
        blanked.writeCharCode(unit == 0x0A ? 0x0A : 0x20);
      }
      if (isFenceMarker) inFence = !inFence;
    } else {
      blanked.write(line);
    }
  }
  return blanked.toString().replaceAllMapped(_inlineCodeRe, (m) => ' ' * m[0]!.length);
}

Iterable<String> _linesInclusive(String input) sync* {
  var start = 0;
  while (start < input.length) {
    final nl = input.indexOf('\n', start);
    if (nl < 0) {
      yield input.substring(start);
      return;
    }
    yield input.substring(start, nl + 1);
    start = nl + 1;
  }
}
